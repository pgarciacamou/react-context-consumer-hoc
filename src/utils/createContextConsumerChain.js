import { consumeContext } from './consumeContext'

// Component that executes the render function with the consumed context
const InnermostComponent = ({ context, render }) => render(context)

// Recursively create a chain of consumers
export function createContextConsumerChain(ContextAPIs) {
  return ContextAPIs.reduce(consumeContext, InnermostComponent)
}
