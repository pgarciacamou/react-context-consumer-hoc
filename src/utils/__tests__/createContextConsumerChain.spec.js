import React from 'react'
import renderer from 'react-test-renderer'
import { createContextConsumerChain } from '../createContextConsumerChain'

describe('createContextConsumerChain', () => {
  it('calls the render function prop with the context', () => {
    const childContextA = { a: 1 }
    const ContextA = React.createContext(childContextA)
    const ConsumerChain = createContextConsumerChain([ContextA])

    const tree = renderer.create(
      <ContextA.Provider value={childContextA}>
        <ConsumerChain
          render={context => {
            expect(context).toEqual(childContextA)
            return (
              <div {...context} />
            )
          }}
        />
      </ContextA.Provider>
    )

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('creates a chain of consumers', () => {
    const childContextA = { a: 1 }
    const childContextB = { b: 1 }
    const ContextA = React.createContext(childContextA)
    const ContextB = React.createContext(childContextB)
    const ConsumerChain = createContextConsumerChain([ContextA, ContextB])

    const tree = renderer.create(
      <ContextA.Provider value={childContextA}>
        <ContextB.Provider value={childContextB}>
          <ConsumerChain
            render={context => {
              expect(context).toEqual({
                ...childContextA,
                ...childContextB
              })
              return (
                <div {...context} />
              )
            }}
          />
        </ContextB.Provider>
      </ContextA.Provider>
    )

    expect(tree.toJSON()).toMatchSnapshot()
  })
})
