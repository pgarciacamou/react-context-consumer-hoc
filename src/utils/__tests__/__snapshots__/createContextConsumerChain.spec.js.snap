// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContextConsumerChain calls the render function prop with the context 1`] = `
<div
  a={1}
/>
`;

exports[`createContextConsumerChain creates a chain of consumers 1`] = `
<div
  a={1}
  b={1}
/>
`;
